### DISCUSSION TITLE
Problems with avalonia.xplat cross compile Desktop project for Linux from Windows.

### DISCUSSION THREAD

**Discussion by sammtan** [AvaloniaUI/Avalonia#15563] [General]
I was trying to `dotnet build` build avalonia.xplat template of Desktop csproj for win-x64, linux-x64, and osx-64. All three of them were compiled successfully, build with .NET Core 8.0.3 SDK. The Windows target just fine, while the macOS target I'm about to ignore for now. I add nothing to the boilerplate code projects, no modification, no native AOT.

For Linux targeting, here is the build command I use.
`dotnet publish -r linux-x64 -c Release --self-contained`

Trying to run it from WSL2 Ubuntu 22, here is the error message I got (in Ubuntu 22 @ WindowsTerminal).
```
Unhandled exception. System.DllNotFoundException: Unable to load shared library 'libICE.so.6' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable:
/mnt/d/blah/AvaX/AvaX.Desktop/bin/Release/net8.0/linux-x64/libICE.so.6: cannot open shared object file: No such file or directory
/mnt/d/blah/AvaX/AvaX.Desktop/bin/Release/net8.0/linux-x64/liblibICE.so.6: cannot open shared object file: No such file or directory
/mnt/d/blah/AvaX/AvaX.Desktop/bin/Release/net8.0/linux-x64/libICE.so.6.so: cannot open shared object file: No such file or directory
/mnt/d/blah/AvaX/AvaX.Desktop/bin/Release/net8.0/linux-x64/liblibICE.so.6.so: cannot open shared object file: No such file or directory

   at Avalonia.X11.ICELib.IceAddConnectionWatch(IntPtr watchProc, IntPtr clientData)
   at Avalonia.X11.X11PlatformLifetimeEvents..ctor(AvaloniaX11Platform platform)
   at Avalonia.X11.AvaloniaX11Platform.Initialize(X11PlatformOptions options)
   at Avalonia.AvaloniaX11PlatformExtensions.<>c.<UseX11>b__0_0()
   at Avalonia.AppBuilder.SetupUnsafe()
   at Avalonia.AppBuilder.Setup()
   at Avalonia.AppBuilder.SetupWithLifetime(IApplicationLifetime lifetime)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, Action`1 lifetimeBuilder)
   at AvaX.Desktop.Program.Main(String[] args) in D:\blah\AvaX\AvaX.Desktop\Program.cs:line 13
Aborted
```
The app won't run. There is a workaround for this, by installing `libice6` and `libsm6`. After done installing (from apt of course), the app run successfully. I can see the window GUI. Well, that's odd. Does every Linux have to first install the two libraries I mentioned earlier, so that this app can run smoothly? What if the Linux installation on which this app runs doesn't have those two libraries? Do you guys know how to overcome this issue, without having the user of my app to manually install `libice6` and `libsm6`?

Anyway, I then tried to add some custom parameters to the csproj file. Here is what it looks like.
`<PackageReference Condition="'$(RuntimeIdentifier)' == 'linux-x64'" Include="Avalonia.X11" Version="$(AvaloniaVersion)" />
`
I'm thinking about adding Avalonia.X11 package reference if -r value of dotnet build command is "linux-x64", but that doesn't work as intended. Kind of make sense, because I think no code from that package has been used in the `Program.cs` (since it's the only file on that csproj), so there's no way it would work.

Any solution for this?

**Comment by maxkatz6** [comment]
Normally, these dependencies are already installed on full Linux distros like Ubuntu. WSL2 isn't exactly GUI-first distro, so it doesn't have many dependencies by default.

**Reply by sammtan** [reply]
Yeah sir, you're right. Tried running this application from my dual boot Ubuntu, and it worked even without having to install both libraries manually, because apparently it is built-in. I guess it's actually just a WSL2 issue. Thank you for your time to answer :D
